src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/queue/QueueRejectionRca.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This RCA reads ThreadPool_RejectionReqs from upstream metrics and it keeps track of
+ * the amount of the time period(tp) when requests in the queues are continuously being rejected.
+ * This RCA is marked as unhealthy if tp we find in write/search queue is above the threshold.
+ */
+public class QueueRejectionRca extends Rca<ResourceFlowUnit> {
+  private static final Logger LOG = LogManager.getLogger(QueueRejectionRca.class);
+  private static final int REJECTION_TIME_PERIOD_IN_SECONDS = 300;
+  private final Metric threadPool_RejectedReqs;
+  private final int rcaPeriod;
+  private final List<QueueRejectedReqsWrapper> queueRejectedReqsWrappers;
+  private final ResourceType resourceType;
+  private int counter;
+  protected Clock clock;
+
+  public <M extends Metric> QueueRejectionRca(final int rcaPeriod, M threadPool_RejectedReqs) {
+    super(5);
+    this.threadPool_RejectedReqs = threadPool_RejectedReqs;
+    this.rcaPeriod = rcaPeriod;
+    resourceType = ResourceType.newBuilder().setThreadpool(ThreadPoolEnum.THREADPOOL_REJECTED_REQS).build();
+    counter = 0;
+    clock = Clock.systemUTC();
+    queueRejectedReqsWrappers = new ArrayList<QueueRejectedReqsWrapper>() {
+      {
+        add(new QueueRejectedReqsWrapper(ThreadPoolType.WRITE));
+        add(new QueueRejectedReqsWrapper(ThreadPoolType.SEARCH));
+      }
+    };
+  }
+
+  @VisibleForTesting
+  public void setClock(Clock clock) {
+    this.clock = clock;
+  }
+
+  @Override
+  public ResourceFlowUnit operate() {
+    counter += 1;
+    long currTimestamp = clock.millis();
+    if (counter == rcaPeriod) {
+      counter = 0;
+      List<TopConsumerSummary> threadpoolSummaries = new ArrayList<>();
+      for (QueueRejectedReqsWrapper wrapper : queueRejectedReqsWrappers) {
+        wrapper.collect(threadPool_RejectedReqs, currTimestamp);
+        // if we've see thread pool rejection in the last 5 mins, the thread pool is considered as contended
+        if (wrapper.hasRejection) {
+          long timePeriod = TimeUnit.MILLISECONDS.toSeconds(currTimestamp - wrapper.getRejectionTimestamp());
+          if (timePeriod >= REJECTION_TIME_PERIOD_IN_SECONDS) {
+            threadpoolSummaries.add(new TopConsumerSummary(wrapper.getThreadPool().toString(), timePeriod));
+          }
+        }
+      }
+      ResourceContext context = null;
+      HotResourceSummary summary = null;
+      if (threadpoolSummaries.isEmpty()) {
+        context = new ResourceContext(Resources.State.HEALTHY);
+      }
+      else {
+        context = new ResourceContext(Resources.State.UNHEALTHY);
+        summary = new HotResourceSummary(this.resourceType,
+            0, threadpoolSummaries.size(), REJECTION_TIME_PERIOD_IN_SECONDS);
+        threadpoolSummaries.sort(
+            (TopConsumerSummary t1, TopConsumerSummary t2) -> Double.compare(t2.getValue(), t1.getValue())
+        );
+        summary.addNestedSummaryList(threadpoolSummaries);
```

Comment:
Wondering if we should have separate top consumers for WRITE and SEARCH.

Reply:
As per our discussion, this is how we planned in the design doc